grakn-core/src/main/java/ai/grakn/concept/ConceptId.java

```diff
@@ -59,15 +60,22 @@ public Object getRawValue(){
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
-
-        ConceptId conceptId1 = (ConceptId) o;
-
-        return getValue().equals(conceptId1.getValue());
+        ConceptId cast = (ConceptId) o;
+        if (conceptId instanceof String) {
```

Comment:
If `conceptId` is graph specific, then indeed the conditional should be changed. And yes, the idea was to avoid doing the toString conversion on non-string types. The comparison on concept ids is called in my inner join and these can easily go in thousands, so added:
a) caching the hashcode
b ) avoiding non-string to string conversion and instead doing a direct comparison